schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,298 @@
+2. CHAMP does not make use, of buckets, nor do common implementations of HAMTs on the JVM (e.g. Clojure, Scala, Java). Storing only entries and links removes the need for an iterative search and compare within buckets and allows direct traversal to the entries required. This is effective for in-memory data structures but is less useful when performing block-by-block traversal with distributed data structures where packing data to reduce traversals may be more important.
+
+### Canonical form
+
+To achieve canonical forms for any given set of `key` / `value` pairs, we note the following properties in the algorithm:
+
+1. We must keep buckets sorted by `key` during both insertion and deletion operations.
+2. We must retain an invariant that states that no non-root node may contain, either directly or via links through child nodes, less than `bucketSize + 1` entries. By applying this strictly during the deletion process, we can generalize that no non-root node without links to child nodes may contain less than `bucketSize + 1` entries. Any non-root node in the tree breaking this rule during the deletion process must have its entries collapsed into a single bucket of `bucketSize` length (i.e. without the entry being removed) and inserted into its parent node in place of the link to the impacted node. We continue to apply this rule recursively up the tree, potentially collapsing additional nodes into their parents.
+
+## Use as a "Set"
+
+The IPLD HashMap can be repurposed as a "Set": a data structure that holds only unique `key`s. Every `value` in a `Set(key, value)` mutation is fixed to some trivial value, such as `true` or `1`. `Has(key)` operations are then simply a `Get(key)` operation that asserts that a value was returned.
+
+## Implementation defaults
+
+Implements need to ship with _sensible defaults_ and be able to create HashMaps without users requiring intimate knowledge of the algorithm and the all of the trade-offs (although such knowledge will help in their optimal use).
+
+These defaults are descriptive rather than prescriptive. New implementations may opt for different defaults, while acknowledging that they will produce different graphs (and therefore CIDs) for the same data as with the defaults listed below. Users may also be provided with facilities to override these defaults to suit their use cases where these defaults do not produce optimal outcomes.
+
+### `hashAlg`
+
+* The default supported hash algorithm for writing IPLD HashMaps is the x64 form of the 128-bit [MurmurHash3](https://github.com/aappleby/smhasher), identified by the multihash name ['murmur3-128'](https://github.com/multiformats/multicodec/blob/master/table.csv). Note the x86 form will produce different output so should not be confused with the x64 form. Additionally, [some JavaScript implementations](https://cimi.io/murmurhash3js-revisited/) do not correctly decompose UTF-8 strings into their constituent bytes for hashing so will not produce portable results.
```

Comment:
The current go implementation relies on the serialization of the two resulting `u64`s, in a specific endianess. This needs to be specced, as it is _not_ part of the original reference implementation. In addition, given that there is no spec it should be clear how edge cases like  https://github.com/spaolacci/murmur3/issues/11  should be handled.

Ref for serialization: https://github.com/spaolacci/murmur3/blob/master/murmur128.go#L52-L61

Reply:
Also the go-hamt-ipld implementation does not use a multihash, but the raw murmur3 hash, this sholud be probably clarified, as the reference to multihash is confusing in this case.

Reply:
@dignifiedquire: re the reference to "multihash" here is simply as a string identifier that goes into the root block under the property "hashAlg", with the intention that any alternative hash algorithm found at "hashAlg" will also be a valid identifier in the multihash table, that's all. e.g. https://github.com/rvagg/iamap/blob/ec2259f5f6108954d511c688aafb91273dc50828/iamap.js#L92-L94 is the only place I use multihash at all in my implementaiton, just a check that it's valid. What I've done is clarified it a bit more at the top of the doc where it first references "multihash" and then put this reference in brackets: _(identified by the multihash name ...)_.

Regarding the specifics of how Go implements multihash and the edge cases, can you help me out with some wording for this? I've had a bunch of trouble with murmur3 across platforms already and it's making me a bit weary of this area.

My go-to JavaScript implementation does some [awkward concat-byte-to-hex work](https://github.com/cimi/murmurhash3js-revisited/blob/de8d6862b5b6507a96c0114547be7944dedd8d86/lib/murmurHash3js.js#L543) which I then have to unpack into bytes again. I have no confidence that the values I'm using will end up being cross platform suitable! I even gave up trying to generate compatible murmur3 hashes in JavaScript as produced by the Scala-to-JavaScript version in [@magik6k's distributed-wiki-search](https://github.com/magik6k/distributed-wiki-search/blob/de58727c3a03476627a9c6eb012d45bf5b4e0e07/client/src/main/scala/WikiSearch.scala#L66) when I was playing with that. Perhaps it's the wrong hash to be defaulting to?

Reply:
Soooo murmur3 is complicated at best, under specified at worst I would argue. But it does work nicely overall and is fast. 

- In regards to compatible versions: I have this rust version https://github.com/dignifiedquire/murmur3/blob/nicer-hashing/src/murmur3_x64_128.rs that has fuzz tests to match against the reference implementation + now matches the bytes of the go implementation for all cases I tried so far. This might be useful to use in JS land as well. 
- If we want to really rely on murmur3 here, I think we should write a spec, for it based on the reference implementation + generate a set of test vectors
  - the minimum here I would say is:
    - define the (byte) order of how the two `u64`s produced by the reference implementation are supposed to be written to bytes
    - say sth like "in case of differences, the reference implementation is to be taken as the correct version, unless a serious bug is discovered" 
- Do we have concrete alternatives in the pipeline we could use, that provide similar speeds?